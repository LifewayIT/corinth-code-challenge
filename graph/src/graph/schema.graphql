enum Gender {
    male
    female
}

scalar DateTime

type Character {
    id: ID!
    birthYear: String!
    eyeColor: String!
    gender: Gender
    hairColor: String
    height: Int!
    homeWorld: Planet
    mass: Int
    name: String!
    skinColor: String
    created: DateTime
    modified: DateTime
    films: [ Film! ]
    species: [ Species! ]
    starships: [ Starship! ]
    picture: String
    url: String
    vehicles: [ Vehicle! ]
}

type Film {
    id: ID!
    name: String!
    picture: String
    episodeId: Int!
    openingCrawl: String
    director: String
    producer: String
    releaseDate: DateTime
    characters: [ Character! ]
    planets: [ Planet! ]
    starships: [ Starship! ]
    vehicles: [ Vehicle! ]
    species: [ Species! ]
    created: DateTime
    edited: DateTime
}

type Planet {
    id: ID!
    name: String!
    picture: String
    rotationPeriodInHours: Int
    orbitalPeriod: Int
    diameter: Int
    climate: String
    gravity: String
    terrain: [String]
    films: [ Film! ]
    surfaceWater: Int
    population: String
    url: String
    created: DateTime
    modified: DateTime
}

type Species {
    id: ID!
    name: String!
    picture: String
    averageHeight: Float
    averageLifespan: Int
    classification: String
    designation: String
    eyeColors: [String!]
    hairColors: [String!]
    homeWorld: Planet
    language: String
    people: [ Character! ]
    films: [ Film! ]
    skinColors: [ String! ]
    created: DateTime
    edited: DateTime
}

type Starship {
    id: ID!
    name: String!
    picture: String
    model: String
    manufacturer: String
    costInCredits: String
    length: Int
    maxAtmospheringSpeed: String
    crew: Int
    passengers: Int
    cargoCapacity: String
    consumables: String
    hyperdriveRating: Float
    MGLT: String
    starshipClass: String
    pilots: [ Character! ]
    films: [ Film! ],
    created: DateTime
    edited: DateTime
}

type Vehicle {
    id: ID!
    name: String!
    picture: String
    cargoCapacity: String
    consumables: String
    costInCredits: String
    crew: Int,
    length: Float
    manufacturer: String
    maxAtmospheringSpeed: Int
    model: String
    passengers: Int
    pilots: [ Character! ]
    films: [ Film! ]
    vehicleClass: String
    created: DateTime
    edited: DateTime
}

union SearchResult = Character | Film | Planet | Species | Starship | Vehicle

type Query {

    search(text: String!, opts: SearchOptionsInput): [SearchResult]

    getCharacter(id: String!): Character
    getFilm(id: String!): Film
    getPlanet(id: String!): Planet
    getSpecies(id: String!): Species
    getStarship(id: String!): Starship
    getVehicle(id: String!): Vehicle

}

input SearchOptionsInput {
    people: Boolean
}

schema {
    query: Query
}
