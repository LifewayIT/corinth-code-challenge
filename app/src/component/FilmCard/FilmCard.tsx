import {Grid} from '@mui/material';
import React, {FunctionComponent} from 'react';
import {FilmService} from '../../graph/film/filmService';
import {BaseCard} from '../BaseCard/BaseCard';
import {FieldDisplay} from '../FieldDisplay/FieldDisplay';

type FilmCardProps = {
    film: FilmService.FilmDetail;
};

export const FilmCard: FunctionComponent<FilmCardProps> = ({film}) => (
    <BaseCard item={film}>
        <Grid container direction="column" spacing={1}>
            <FieldDisplay label="Episode ID" value={film.episodeId}/>
            <FieldDisplay label="Director" value={film.director}/>
            <FieldDisplay label="Producer" value={film.producer}/>
            <FieldDisplay label="Release Date" value={film.releaseDate as string}/>
            <FieldDisplay label="Opening Crawl" value={film.openingCrawl}/>
        </Grid>
    </BaseCard>
);
