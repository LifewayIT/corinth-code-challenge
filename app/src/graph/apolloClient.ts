import {ApolloClient, ApolloLink, createHttpLink, InMemoryCache, NormalizedCacheObject} from '@apollo/client';
import {onError} from '@apollo/client/link/error';
import {GraphQLError} from 'graphql';
import {Config} from '../util/config';
import introspection from './introspection';

const requestLink = createHttpLink({
    uri: Config.getGraphQLEndpoint()
});

type AppSyncError = GraphQLError & {
    errorType: string;
};

const errorLink = onError(
    ({graphQLErrors, networkError}) => {
        if (graphQLErrors) {
            (graphQLErrors as AppSyncError[]).forEach(() => {
                // TODO: handle specific errors
            });
        } else if (networkError) {
            // TODO: do something
        }
    }
);

export const apolloCache = new InMemoryCache({
    possibleTypes: introspection.possibleTypes,
    addTypename: true
});

// assemble and export the client
export const apolloClient = new ApolloClient<NormalizedCacheObject>({
    cache: apolloCache,
    link: ApolloLink.from([
        errorLink,
        requestLink
    ]),
    defaultOptions: {}
});
