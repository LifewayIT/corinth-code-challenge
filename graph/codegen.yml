overwrite: true
schema: ./src/graph/schema.graphql
generates:
  # creates typescript definitions
  src/graph/generated.ts:
    plugins:
      - typescript
      - typescript-resolvers
