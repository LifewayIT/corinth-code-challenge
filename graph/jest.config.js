global.environment = {
    production: false
};

module.exports = {
    moduleFileExtensions: [
        'ts',
        'tsx',
        'js'
    ],
    setupFiles: [
        '<rootDir>/jest.setup.ts'
    ],
    transform: {
        '\\.css$': 'jest-raw-loader',
        '\\.sql$': 'jest-raw-loader',
        '\\.txt$': 'jest-raw-loader',
        '^.+\\.tsx?$': 'ts-jest'
    },
    coverageThreshold: {
        global: {
            branches: 90,
            functions: 90,
            lines: 90,
            statements: 90
        }
    }
};
