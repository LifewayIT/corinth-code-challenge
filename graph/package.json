{
  "name": "corinth-code-challenge-graph",
  "version": "1.0.0",
  "description": "",
  "main": "dist/server.js",
  "scripts": {
    "build": "npm run generate:graphql && npm run lint && webpack",
    "generate:graphql": "gql-gen --config codegen.yml",
    "lint": "eslint src --ext .ts",
    "start": "npm run build && node dist/server.js",
    "test": "jest"
  },
  "author": {
    "name": "Robin Schultz",
    "email": "violinrobin@gmail.com",
    "url": "https://github.com/anigenero"
  },
  "license": "UNLICENSED",
  "private": true,
  "dependencies": {
    "apollo-server": "^3.3.0",
    "axios": "^0.21.4",
    "graphql": "^15.6.0",
    "graphql-parse-resolve-info": "^4.12.0",
    "luxon": "^2.0.2",
    "rxjs": "^7.3.0"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.15.6",
    "@graphql-codegen/cli": "^2.1.1",
    "@graphql-codegen/typescript": "^2.1.0",
    "@graphql-codegen/typescript-resolvers": "^2.2.1",
    "@types/bluebird": "^3.5.36",
    "@types/jest": "^27.0.2",
    "@types/luxon": "^2.0.4",
    "@types/mongodb": "^4.0.7",
    "@typescript-eslint/eslint-plugin": "^4.31.2",
    "@typescript-eslint/parser": "^4.31.2",
    "babel-loader": "^8.2.2",
    "bluebird": "^3.7.2",
    "core-js": "^3.18.0",
    "eslint": "^7.32.0",
    "eslint-plugin-import": "^2.24.2",
    "eslint-plugin-prefer-arrow": "^1.2.3",
    "jest": "^27.2.1",
    "jest-cli": "^27.2.1",
    "jest-raw-loader": "^1.0.1",
    "json-schema-to-typescript": "^10.1.5",
    "ts-jest": "^27.0.5",
    "ts-loader": "^9.2.6",
    "typescript": "^4.4.3",
    "webpack-cli": "^4.8.0",
    "webpack-node-externals": "^3.0.0"
  }
}
