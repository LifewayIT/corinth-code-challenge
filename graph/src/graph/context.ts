import {SWAPIClient} from '../swClient';

export type SWContext = {
    client: SWAPIClient;
};
