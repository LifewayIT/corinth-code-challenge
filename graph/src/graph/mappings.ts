import {Character, Film, Gender, Planet, Species, Starship, Vehicle} from './generated';

export const getIdFromUrl = (url: SWAPI.Url): string =>
    url.substr(0, url.length - 1).split('/').pop();

export const toFloat = (str: string): number => {
    const result = parseFloat(str);
    return !isNaN(result) ? result : null;
};

export const toInt = (str: string): number => {
    const result = parseInt(str, 10);
    return !isNaN(result) ? result : null;
};

export const toList = (str: string, delimiter = ','): string[] => {
    if (str) {
        return str.split(delimiter).map((val) => val.trim());
    } else {
        return [];
    }
};

export const toGender = (str: string): Gender => {
    switch (str) {
        case 'male':
            return Gender.Male;
        case 'female':
            return Gender.Female;
        default:
            return null;
    }
};

export const toPictureAsset = (id: string, asset: 'characters' | 'films' | 'planets' | 'species' | 'starships' | 'vehicles'): string =>
    `https://starwars-visualguide.com/assets/img/${asset}/${id}.jpg`;

export const toCharacter = (character: SWAPI.Person): Character => {
    const id = getIdFromUrl(character.url);
    return {
        __typename: 'Character',
        id,
        picture: toPictureAsset(id, 'characters'),
        birthYear: character.birth_year,
        eyeColor: character.eye_color,
        gender: toGender(character.gender),
        height: toInt(character.height),
        hairColor: character.hair_color,
        mass: toInt(character.mass),
        skinColor: character.skin_color,
        name: character.name,
        created: character.created,
        modified: character.edited
    };
};

export const toPlanet = (planet: SWAPI.Planet): Planet => {
    const id = getIdFromUrl(planet.url);
    return {
        __typename: 'Planet',
        id,
        name: planet.name,
        picture: toPictureAsset(id, 'planets'),
        rotationPeriodInHours: toInt(planet.rotation_period),
        orbitalPeriod: toInt(planet.orbital_period),
        diameter: toInt(planet.diameter),
        climate: planet.climate,
        gravity: planet.gravity,
        terrain: toList(planet.terrain),
        surfaceWater: toInt(planet.surface_water),
        population: planet.population,
        created: planet.created,
        modified: planet.edited
    };
};

export const toFilm = (film: SWAPI.Film): Film => {
    const id = getIdFromUrl(film.url);
    return {
        __typename: 'Film',
        id,
        name: film.title,
        picture: toPictureAsset(id, 'films'),
        releaseDate: film.release_date,
        director: film.director,
        episodeId: film.episode_id,
        openingCrawl: film.opening_crawl,
        producer: film.producer,
        created: film.created,
        edited: film.edited
    };
};

export const toStarship = (starship: SWAPI.Starship): Starship => {
    const id = getIdFromUrl(starship.url);
    return {
        __typename: 'Starship',
        id,
        name: starship.name,
        picture: toPictureAsset(id, 'starships'),
        model: starship.model,
        manufacturer: starship.manufacturer,
        costInCredits: starship.cost_in_credits,
        length: toInt(starship.length),
        maxAtmospheringSpeed: starship.max_atmosphering_speed,
        crew: toInt(starship.crew),
        passengers: toInt(starship.crew),
        cargoCapacity: starship.cargo_capacity,
        consumables: starship.consumables,
        hyperdriveRating: toFloat(starship.hyperdrive_rating),
        MGLT: starship.MGLT,
        starshipClass: starship.starship_class,
        created: starship.created,
        edited: starship.edited
    };
};

export const toSpecies = (species: SWAPI.Species): Species => {
    const id = getIdFromUrl(species.url);
    return {
        __typename: 'Species',
        id,
        name: species.name,
        picture: toPictureAsset(id, 'species'),
        averageHeight: toFloat(species.average_height),
        averageLifespan: toInt(species.average_lifespan),
        classification: species.classification,
        designation: species.designation,
        eyeColors: toList(species.eye_colors),
        hairColors: toList(species.hair_colors),
        language: species.language,
        skinColors: toList(species.skin_colors),
        created: species.created,
        edited: species.edited
    };
};

export const toVehicle = (vehicle: SWAPI.Vehicle): Vehicle => {
    const id = getIdFromUrl(vehicle.url);
    return {
        __typename: 'Vehicle',
        id,
        name: vehicle.name,
        picture: toPictureAsset(id, 'vehicles'),
        cargoCapacity: vehicle.cargo_capacity,
        consumables: vehicle.consumables,
        costInCredits: vehicle.cost_in_credits,
        crew: toInt(vehicle.crew),
        length: toInt(vehicle.length),
        manufacturer: vehicle.manufacturer,
        maxAtmospheringSpeed: toInt(vehicle.max_atmosphering_speed),
        model: vehicle.model,
        passengers: toInt(vehicle.passengers),
        vehicleClass: vehicle.vehicle_class,
        created: vehicle.created,
        edited: vehicle.edited
    };
};
