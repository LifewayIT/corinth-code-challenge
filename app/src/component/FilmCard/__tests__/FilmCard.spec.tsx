import React from 'react';
import {create} from 'react-test-renderer';
import {Film} from '../../../graph/generated';
import {FilmCard} from '../FilmCard';

describe('FilmCard', () => {

    it('should render correctly', () => {

        const film: Film = {
            id: '3',
            name: 'Return of the Jedi',
            episodeId: 6,
            openingCrawl: 'Luke Skywalker has returned to\r\nhis home planet of Tatooine in',
            director: 'Richard Marquand',
            producer: 'Howard G. Kazanjian, George Lucas, Rick McCallum',
            releaseDate: '1983-05-25',
            created: '2014-12-18T10:39:33.255000Z',
            edited: '2014-12-20T09:48:37.462000Z'
        };

        const tree = create(
            <FilmCard film={film}/>
        ).toJSON();

        expect(tree).toMatchSnapshot();

    });

});
