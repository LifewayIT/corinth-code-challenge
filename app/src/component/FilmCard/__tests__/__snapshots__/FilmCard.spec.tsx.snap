// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmCard should render correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1902mtl-MuiPaper-root-MuiCard-root"
>
  <img
    alt="Return of the Jedi"
    className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
    height="500"
  />
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
    >
      Return of the Jedi
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-ffjoah-MuiGrid-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-1744jwa-MuiGrid-root"
          >
            <strong>
              Episode ID
              :
            </strong>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
          >
            6
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-1744jwa-MuiGrid-root"
          >
            <strong>
              Director
              :
            </strong>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
          >
            Richard Marquand
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-1744jwa-MuiGrid-root"
          >
            <strong>
              Producer
              :
            </strong>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
          >
            Howard G. Kazanjian, George Lucas, Rick McCallum
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-1744jwa-MuiGrid-root"
          >
            <strong>
              Release Date
              :
            </strong>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
          >
            1983-05-25
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-11lq3yg-MuiGrid-root"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-1744jwa-MuiGrid-root"
          >
            <strong>
              Opening Crawl
              :
            </strong>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
          >
            Luke Skywalker has returned to
his home planet of Tatooine in
          </div>
        </div>
      </div>
    </p>
  </div>
</div>
`;
